### DISCUSSION TITLE
How to achieve fast queries with low-frequency inserts and high QPS?

### DISCUSSION THREAD

**Discussion by schuberttobias** [milvus-io/milvus#16867] [Q&A and General discussion]
Hi Milvus team!

We want to achieve fast queries with low-frequency inserts and high QPS.

Currently there is high query time and CPU saturation at 45 %. After having read the documentation and experimenting with different configuration parameters (e.g. `dataCoord.segment.maxSize`), we can't find a solution and kindly ask for your help.

### Collection
  - 4 fields
    - primary_id (INT64)
    - vector_data (FLOAT_VECTOR)
    - rdbms_id1 (INT64)
    - rdbms_id2 (INT64)
  - 2 shards (default)

### Configuration
- AWS EC2 c5a.4xlarge. 16 cores, 32 GB RAM. 100 GB disk.
- Ubuntu Server 18.04.
- Milvus Standalone 2.0.2-1 from Launchpad PPA.
- `milvus.yaml`. Related parameters:
  - `dataCoord.segment.maxSize`: 128
  - `rootCoord.minSegmentSizeToEnableIndex`: 1024 (default)
  - `dataCoord.segment.sealProportion`: 0.75 (default)
  - Full config: https://gist.github.com/schuberttobias/cef2832eeb0bbbe23bf6178234cf59a6
- Index
  - IVF_SQ8
  - nlist=128
  - metric_type=IP
- Query parameters
  - nprobe=16
  - metric_type=IP

## Goals
- Vectors: up to 50 million. Dimension: 128.
- Inserts
  - 400 vectors inserted every 7 seconds. I.e. ~200.000 vectors per hour.
- Query
  - 150 Python processes execute 1.66 QPS each: 250 QPS total.
  - Max. time per query: 400ms, ideally less.
  - TOP_K=15.
  - output_fields: rdbms_id1, rdbms_id2 (see above)

### 1. High query time
In the first 10 hours query time is okay:
1. In the beginning, query time is low (<100ms for 50,000 vectors) - which is great.
2. Number of vectors increases, segment grows, more parts of query are brute force and take longer, but always less than 1000ms.
3. Every 2 hours max. segment size 128 is reached. Index is built, Query is fast again.

However after 10 hours, with approximately 2 million vectors in Milvus, the brute force search becomes as slow as 3000-5000ms per query. Also, the CPU is saturated and never exceeds 45 %. There is a direct relation between CPU saturation and high query time.

#### Screenshot CPU usage saturated
![image](https://user-images.githubusercontent.com/17709711/167446934-7f5a64d1-447f-4cdb-bf68-f0a73f370cbb.png)
(x-axis=time, y-axis=CPU usage in percent of Milvus standalone running on Ubuntu)

### 2. Relation between segment size and index build interval
- Changing max. segment size to 512 (= **4** * 128), leads to index being built every 8 hours (= **4** * 2 hours). So there is a direct relation between index build time and segment size.

## Logs
- Logs from time frame when CPU was saturated at 45 % and query time exceeded 3000ms.
- Index built at approx. `13:39:00 +00:00`:
```
standalone-9-2022-04-18T13-42-09.454.log:[2022/04/18 13:39:34.872 +00:00] [DEBUG] [root_coord.go:856] ["start build index"] ["index name"=_default_idx] ["field name"=fingerprint] ["segment id"=432611237475385346]
standalone-9-2022-04-18T13-42-09.454.log:[2022/04/18 13:39:37.135 +00:00] [DEBUG] [time_recorder.go:78] ["IndexBuildTask 432613405457645570: build index done (0ms)"]
standalone-9-2022-04-18T13-42-09.454.log:[2022/04/18 13:39:43.285 +00:00] [DEBUG] [root_coord.go:856] ["start build index"] ["index name"=_default_idx] ["field name"=fingerprint] ["segment id"=432611237475385345]
standalone-9-2022-04-18T13-42-09.454.log:[2022/04/18 13:39:46.136 +00:00] [DEBUG] [time_recorder.go:78] ["IndexBuildTask 432613407659655170: build index done (0ms)"] 
```
- Query time is high before building index. 10 minutes after "build index done" query time is low again.
```
standalone-9-2022-04-18T13-20-29.261.log:[2022/04/18 13:13:13.874 +00:00] [DEBUG] [time_recorder.go:78] ["searchTask PostExecute: done (3228ms)"]
...
standalone-9-2022-04-18T13-36-03.885.log:[2022/04/18 13:35:34.175 +00:00] [DEBUG] [time_recorder.go:78] ["searchTask PostExecute: done (3499ms)"]
...
standalone-9-2022-04-18T13-42-09.454.log:[2022/04/18 13:39:34.033 +00:00] [DEBUG] [time_recorder.go:78] ["searchTask PostExecute: done (3545ms)"]
...
standalone-9-2022-04-18T13-42-09.454.log:[2022/04/18 13:39:52.085 +00:00] [DEBUG] [time_recorder.go:78] ["searchTask PostExecute: done (2205ms)"]
...
standalone-9-2022-04-18T13-42-09.454.log:[2022/04/18 13:40:02.556 +00:00] [DEBUG] [time_recorder.go:78] ["searchTask PostExecute: done (2122ms)"]
...
standalone-9-2022-04-18T13-42-09.454.log:[2022/04/18 13:40:25.578 +00:00] [DEBUG] [time_recorder.go:78] ["searchTask PostExecute: done (2119ms)"]
...
standalone-9-2022-04-18T13-47-02.559.log:[2022/04/18 13:42:51.826 +00:00] [DEBUG] [time_recorder.go:78] ["searchTask PostExecute: done (2152ms)"]
...
standalone-9-2022-04-18T13-52-23.429.log:[2022/04/18 13:50:09.196 +00:00] [DEBUG] [time_recorder.go:78] ["searchTask PostExecute: done (1950ms)"]
...
standalone-9-2022-04-18T13-52-23.429.log:[2022/04/18 13:50:32.348 +00:00] [DEBUG] [time_recorder.go:78] ["searchTask PostExecute: done (977ms)"]
...
standalone-9-2022-04-18T13-52-23.429.log:[2022/04/18 13:50:39.101 +00:00] [DEBUG] [time_recorder.go:78] ["searchTask PostExecute: done (171ms)"]
...
standalone-9-2022-04-18T13-52-23.429.log:[2022/04/18 13:51:19.993 +00:00] [DEBUG] [time_recorder.go:78] ["searchTask PostExecute: done (63ms)"]
standalone-9-2022-04-18T13-52-23.429.log:[2022/04/18 13:51:20.008 +00:00] [DEBUG] [time_recorder.go:78] ["searchTask PostExecute: done (67ms)"]
standalone-9-2022-04-18T13-52-23.429.log:[2022/04/18 13:51:20.024 +00:00] [DEBUG] [time_recorder.go:78] ["searchTask PostExecute: done (55ms)"]
standalone-9-2022-04-18T13-52-23.429.log:[2022/04/18 13:51:20.041 +00:00] [DEBUG] [time_recorder.go:78] ["searchTask PostExecute: done (71ms)"]
standalone-9-2022-04-18T13-52-23.429.log:[2022/04/18 13:51:20.057 +00:00] [DEBUG] [time_recorder.go:78] ["searchTask PostExecute: done (67ms)"]
standalone-9-2022-04-18T13-52-23.429.log:[2022/04/18 13:51:20.073 +00:00] [DEBUG] [time_recorder.go:78] ["searchTask PostExecute: done (82ms)"]
standalone-9-2022-04-18T13-52-23.429.log:[2022/04/18 13:51:20.089 +00:00] [DEBUG] [time_recorder.go:78] ["searchTask PostExecute: done (83ms)"]
standalone-9-2022-04-18T13-52-23.429.log:[2022/04/18 13:51:20.107 +00:00] [DEBUG] [time_recorder.go:78] ["searchTask PostExecute: done (101ms)"]
standalone-9-2022-04-18T13-52-23.429.log:[2022/04/18 13:51:20.124 +00:00] [DEBUG] [time_recorder.go:78] ["searchTask PostExecute: done (106ms)"]
standalone-9-2022-04-18T13-52-23.429.log:[2022/04/18 13:51:20.139 +00:00] [DEBUG] [time_recorder.go:78] ["searchTask PostExecute: done (85ms)"]
standalone-9-2022-04-18T13-52-23.429.log:[2022/04/18 13:51:20.155 +00:00] [DEBUG] [time_recorder.go:78] ["searchTask PostExecute: done (43ms)"]
standalone-9-2022-04-18T13-52-23.429.log:[2022/04/18 13:51:20.171 +00:00] [DEBUG] [time_recorder.go:78] ["searchTask PostExecute: done (55ms)"]
standalone-9-2022-04-18T13-52-23.429.log:[2022/04/18 13:51:20.188 +00:00] [DEBUG] [time_recorder.go:78] ["searchTask PostExecute: done (32ms)"]
standalone-9-2022-04-18T13-52-23.429.log:[2022/04/18 13:51:20.203 +00:00] [DEBUG] [time_recorder.go:78] ["searchTask PostExecute: done (34ms)"]
standalone-9-2022-04-18T13-52-23.429.log:[2022/04/18 13:51:20.220 +00:00] [DEBUG] [time_recorder.go:78] ["searchTask PostExecute: done (48ms)"]
standalone-9-2022-04-18T13-52-23.429.log:[2022/04/18 13:51:20.237 +00:00] [DEBUG] [time_recorder.go:78] ["searchTask PostExecute: done (54ms)"]
standalone-9-2022-04-18T13-52-23.429.log:[2022/04/18 13:51:20.252 +00:00] [DEBUG] [time_recorder.go:78] ["searchTask PostExecute: done (32ms)"]
```

## Questions
1. Why does Milvus not use more than 50% of CPU?
    1. Is the cause that there is only 1 query node? 
    2. If yes: How can we enable additional query nodes in Milvus standalone?
2. How can we achieve <400ms per query with 250 QPS and 50 million vectors?

**Comment by cydrain** [comment]
Test index IVF_SQ8 with following parameters:
 - nlist = 1024
 - nprobe = 16
 - topk = 10

*Scenario 1*:
Run **knowhere** only, keep querying with different NQs.
Whatever NQ is, CPU load all reach 100%.

*Scenario 2*:
Run **Milvus standalone**, keep query with different NQs.
When NQ <= 50, CPU load is about 10%; when NQ >= 100, CPU load increases with NQ.

nq = 10
![1-10](https://user-images.githubusercontent.com/3992404/168104711-05ba96b9-9992-4e3a-8346-1817fc337565.png)

nq = 20
![1-20](https://user-images.githubusercontent.com/3992404/168104868-7f895409-a0b7-4b76-9b2c-14f46e62073c.png)

nq = 50
![1-50](https://user-images.githubusercontent.com/3992404/168105007-94feb363-4193-49c1-9d02-933ced6ddd30.png)

nq = 100
![1-100](https://user-images.githubusercontent.com/3992404/168105061-72c027ec-1541-4367-91b1-d1d94f7d259b.png)

nq = 200
![1-200](https://user-images.githubusercontent.com/3992404/168105114-6b9b0cc2-6ec4-4204-993e-889ecbe20099.png)

nq = 500
![1-500](https://user-images.githubusercontent.com/3992404/168105159-a2caf679-e806-4d30-a138-205044717f98.png)

nq = 1000
![1-1000](https://user-images.githubusercontent.com/3992404/168105216-dc5e257b-50d7-4aae-94ce-0728ef91b82c.png)

**Comment by cydrain** [comment]
For your questions:

#### 1. Why does Milvus not use more than 50% of CPU?
#### i. Is the cause that there is only 1 query node?
	It's because there is only 1 Proxy and 1 Querynode inside Milvus standalone.
    Both Proxy and Querynode handle search request serially, because there is system overhead, CPU load cannot reach 100%.
#### ii. If yes: How can we enable additional query nodes in Milvus standalone?
	No, standalone has only one query node inside, we cannot enable additional query nodes.
#### 2. How can we achieve <400ms per query with 250 QPS and 50 million vectors?
Why not try Milvus cluster ?

**Reply by xiaofan-luan** [reply]
In Milvus 2.1 we fixed huge amount of issues on performance, including:
1) Search combination
2) fix some perf bottleneck in reduce
3) change the schedule algorithm so the CPU utility could reach higher
4) Enhance attr filtering.
achieve <400ms per query with 250 QPS and 50 million vectors is kind of hard on single machine
Things to do:
1) wait 2.1 release and change to 2.1
2) use HNSW index rather than Faiss IVF
3) Use larger batch instead nq=1
4) Scale out/up

**Reply by schuberttobias** [reply]
@cydrain @xiaofan-luan : Thank you for your help! We are currently experimenting with all the things you suggested (e.g., Milvus Cluster, scaling proxy and query nodes, different nq, HNSW index, etc.). We will report back when we have a conclusion!

**Reply by schuberttobias** [reply]
@cydrain @xiaofan-luan 

Thanks again for your help!

We were able to apply most of your tips and I am reporting back our findings:

- tested against Milvus Cluster 2.0.2-1 using docker-compose on c5.4xlarge
- nlist/nprobe/nq
  - Changing from nlist=128,nprobe=16,nq=5 to nlist=2048,nprobe=16,nq=5 increased search performance by factor 2,66.
- scaling
  - With few vectors (300.000) in Milvus: the default single proxy node is the bottleneck. Increasing to 2 proxy nodes increases search performance by factor 2. Increasing query nodes brings no benefits.
  - With many vectors (25 million) in Milvus: the default single query node is the bottleneck. Because of the many vectors, it fully utilizes the CPU. Adding a second proxy or query node does not increase performance.
- HNSW
  - switching from IVF to HNSW decreased performance. Limitation: we used randomly generated vector data, not real application data.
- MetricType
  - switching from IP to L2 decreased performance.

With scaling out (3x proxy, 3x query node), adapting nlist (2048) we were able to reach **~40 QPS** with nq=5,nprobe=16 and 20 million vectors in Milvus, This currently suffices for our use case. Before optimization it was <= 15 QPS and too low.

We are looking forward to upgrading to Milvus v2.2 and seeing if it brings further performance benefits for our use case.

Mention @smeikalr
